feat: map SAML role to a Frappe user role or role profile

Alchez opened the pull request:
Closes #5.

https://github.com/user-attachments/assets/2ab4a06f-4ce6-4ff3-ad2a-2402e1a56d12

Commit:
feat: map SAML role to a Frappe user role

agritheory commented:
- [ ] Add a mapping to the fixtures and matching and mismatching asserts on roles
- [x] Detect ports and inject those into the realms config
- [ ] Add a small selection of Roles
- [x] Add a setup function of create user(s) and SAML Login key config
- [ ] Strategically (un) assign roles to users for testing (Sys Manager, Customer / Supplier)
- [x] Add option to select Role or Role Profile in SAML Key Mapping

agritheory committed:
feat: role profile integration

Commit:
test: detect bench port to generate realm export file

Commit:
test: add test setup function

Commit:
test: apply SAML roles using new toggle

Commit:
ci: setup bench services to generate test data

Alchez commented:
@agritheory I added some commits that checks off some of the items you mentioned. Can you check test this out and see if it works for you locally?

Commit:
fix: update Keycloak test roles

agritheory commented:
What directory are you starting the docker compose from to collect the ports? I tried `docker compose down && docker compose up --build` from `~/frappe-bench/apps/saml/saml/tests`

```
realm-processor-1  | Processing realm export with base URL: http://localhost:8000
realm-processor-1  | Found frappe-saml client, updating URLs to http://localhost:8000
realm-processor-1  | Traceback (most recent call last):
realm-processor-1  |   File "/app/process_realm.py", line 41, in <module>
realm-processor-1  |     main()
realm-processor-1  |   File "/app/process_realm.py", line 34, in main
realm-processor-1  |     with open(output_file, "w") as f:
realm-processor-1  |          ^^^^^^^^^^^^^^^^^^^^^^
realm-processor-1  | IsADirectoryError: [Errno 21] Is a directory: '/app/realm-export.json'
```

agritheory committed:
test: add role profile fixture

agritheory commented:
Copied from #5: 

SAML also supports groups, so I think we can update this with a clearer mapping API:

- SAML Group to Frappe Role Profile
- SAML Role to Frappe Role
- Groups/ Role Profiles will unset / override any Role-specific mappings

The code for this is mostly there, I think we can rename the "SAML Role" column to "SAML Group or Role" in SAML Group Mapping. 

I added a Role Profile fixture to tests/data.

Alchez commented:
> What directory are you starting the docker compose from to collect the ports? I tried `docker compose down && docker compose up --build` from `~/frappe-bench/apps/saml/saml/tests`
> 
> ```
> realm-processor-1  | Processing realm export with base URL: http://localhost:8000
> realm-processor-1  | Found frappe-saml client, updating URLs to http://localhost:8000
> realm-processor-1  | Traceback (most recent call last):
> realm-processor-1  |   File "/app/process_realm.py", line 41, in <module>
> realm-processor-1  |     main()
> realm-processor-1  |   File "/app/process_realm.py", line 34, in main
> realm-processor-1  |     with open(output_file, "w") as f:
> realm-processor-1  |          ^^^^^^^^^^^^^^^^^^^^^^
> realm-processor-1  | IsADirectoryError: [Errno 21] Is a directory: '/app/realm-export.json'
> ```

@agritheory I forgot to mention this in the PR, but I've updated the run steps in the `readme` file. I've added a new script that should generate the necessary files considering your current bench port and then spawning the docker containers.

Basically you want to run the following command inside the `tests` folder:

```sh
./keycloak.sh
```

If you want to add a build step, then you could also do:

```sh
./keycloak.sh --build
```

---

> Copied from #5:
> 
> SAML also supports groups, so I think we can update this with a clearer mapping API:
> 
> * SAML Group to Frappe Role Profile
> * SAML Role to Frappe Role
> * Groups/ Role Profiles will unset / override any Role-specific mappings
> 

I did the following steps:
- Added a "Group" in the Keycloak IdP (Keycloak defines this as a "set of attributes and role mappings that can be applied to a user")
- Added some roles in the `frappe` realm inside Keycloak
- Assigned those roles to the new role Group
- Assigned the SAML user to the Group
- When logging in, the SAML auth response returns only the Group's roles, and not information about the Group itself.

Now, there are ways to configure an IdP to return custom attributes in the auth response, but we would have to do that as an additional manual step.

What do you think?

---

> I added a Role Profile fixture to tests/data.

I had actually deleted this file, since it was (is?) causing an issue during insertion. There seems to be some changes with background document methods and file locks that I'm unable to solve. If you run a `bench reinstall` using that JSON, do you get a file-lock error?
